pages/blog/graphql-error-handling-with-fp.mdx

```diff
@@ -0,0 +1,631 @@
+- return `null`: this is not satisfactory because we can't know the reason the operation returned `null`
+
+Let's make the following modification to handle all possible paths:
+
+```graphql
+interface BaseError {
+  message: String!
+}
+
+type InvalidInputError implements BaseError {
+  message: String!
+}
+
+type NotFoundError implements BaseError {
+  message: String!
+}
+
+type UnknownError implements BaseError {
+  message: String!
+}
+
+type NotAllowedError implements BaseError {
+  message: String!
+}
+
+type Entity {
+  id: ID!
+  name: String!
+}
+
+union EntityResult =
+  | Entity
+  | NotFoundError
+  | NotAllowedError
+  | InvalidInputError
+  | UnknownError
+
+type Query {
+  entity(id: ID!, userId: ID!): EntityResult!
+}
+```
+
+We define a `BaseError` interface that all concrete Errors implement. And we define the query result as a Union type of the happy path (an `Entity`) and the other possible results (`NotFoundError`, `NotAllowedError`, etc.). In this manner we extensively describe all possible results, with the added benefit that we can make our result type non-nullable (with the `!` character).
+
+For a more detailed explanation, you can read this article from Sasha Solomon: [200 OK! Error Handling in GraphQL](https://sachee.medium.com/200-ok-error-handling-in-graphql-7ec869aec9bc).
+
+## Handle those Schema changes into the GraphQL server
+
+Since we changed the Schema, we also need to change the query, mutation and types resolvers.
+
+For simple cases, we can just return the correct object from the query resolver:
+
+```typescript
+const resolvers = {
+  Query: {
+    entity: async (parent, { id, userId }, context) => {
+      const entity = await context.api.getEntityById(id);
+      const isAllowed = await context.api.isUserAllowedForEntity(id, userId);
+
+      if (!isAllowed) {
+        return {
+          __typename: 'NotAllowedError',
+          message: `User ${userId} not allowed for entity ${id}`,
+        };
+      }
+      if (!entity) {
+        return {
+          __typename: 'NotFoundError',
+          message: `Entity ${id} not found`,
+        };
+      }
+      return {
+        __typename: 'Entity',
+        ...entity,
+      };
+    },
+  },
+};
+```
+
+Since we include the `__typename` and the fields match the Schema type definitions, there is no need to define type resolvers.
+
+For more details on this solution, you can read this article from [Laurin Quast](https://github.com/n1ru4l): [Handling GraphQL errors like a champ with unions and interfaces](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/).
+
+
+# Typesafe error handling
+
+
+Let's now take it to the next level using Typescript and GraphQL Code Generator.
+
+[Typescript](https://www.typescriptlang.org/) is a superset of Javascript which adds type annotations to Javascript code and therefore type safety (with caveats) at compile-time.
```

Comment:
We can assume that the audience knows TypeScript and remove this paragraph.

(Last State of JS survey shows that most people know it or are using it: https://2021.stateofjs.com/en-US/other-tools/#javascript_flavors